pallets/order-book/src/lib.rs

```diff
@@ -0,0 +1,639 @@
+			// might need decimals from currency, but should hopefully be able to use FP
+			// price/amounts from FP balance
+			T::AssetRegistry::metadata(&order.asset_out_id).ok_or(Error::<T>::InvalidAssetId)?;
+
+			T::AssetRegistry::metadata(&order.asset_in_id).ok_or(Error::<T>::InvalidAssetId)?;
+
+			let sell_amount = order.buy_amount.ensure_mul(order.price)?;
+
+			ensure!(
+				T::TradeableAsset::can_reserve(order.asset_out_id, &account_id, sell_amount),
+				Error::<T>::InsufficientAssetFunds,
+			);
+			T::TradeableAsset::unreserve(
+				order.asset_out_id,
+				&order.placing_account,
+				order.max_sell_amount,
+			);
+			T::ReserveCurrency::unreserve(
+				&order.placing_account,
+				T::Fees::fee_value(T::OrderFeeKey::get()),
+			);
+			T::TradeableAsset::transfer(
+				order.asset_in_id,
+				&account_id,
+				&order.placing_account,
+				order.buy_amount,
+			)?;
+			T::TradeableAsset::transfer(
+				order.asset_out_id,
+				&order.placing_account,
+				&account_id,
+				sell_amount,
+			)?;
+			Self::remove_order(order.order_id)?;
+			Self::deposit_event(Event::OrderFulfillment {
+				order_id,
+				placing_account: order.placing_account,
+				fulfilling_account: account_id,
+				partial_fulfillment: true,
+				currency_in: order.asset_in_id,
+				currency_out: order.asset_out_id,
+				fulfillment_amount: order.buy_amount,
+				sell_price_limit: order.price,
+			});
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Get all orders for an account
+		/// Provided for frontend to grab orders for an individual account
+		pub fn get_account_orders(
+			account_id: T::AccountId,
+		) -> Result<
+			sp_std::vec::Vec<(
+				T::Hash,
+				Order<T::Hash, T::AccountId, T::AssetCurrencyId, T::ForeignCurrencyBalance>,
+			)>,
+			Error<T>,
+		> {
+			Ok(<UserOrders<T>>::iter_prefix(account_id).collect())
+		}
+
+		/// Get all orders
+		/// Provided for frontend to grab all open orders
+		pub fn get_all_orders() -> Result<
+			sp_std::vec::Vec<(
+				T::Hash,
+				Order<T::Hash, T::AccountId, T::AssetCurrencyId, T::ForeignCurrencyBalance>,
+			)>,
+			Error<T>,
+		> {
+			Ok(<Orders<T>>::iter().collect())
+		}
+
+		/// Remove an order from storage
+		pub fn remove_order(order_id: T::Hash) -> DispatchResult {
+			let order = <Orders<T>>::get(order_id)?;
+			<UserOrders<T>>::remove(&order.placing_account, order.order_id);
+			<Orders<T>>::remove(order.order_id);
+			let mut orders = <AssetPairOrders<T>>::get(order.asset_in_id, order.asset_out_id);
+			orders.retain(|o| *o != order.order_id);
+			<AssetPairOrders<T>>::insert(order.asset_in_id, order.asset_out_id, orders);
+			Ok(())
+		}
+
+		/// Generate a hash to be used for the order id using the placing
+		/// account, asset IDs, and pallet nonce.
+		pub fn gen_hash(
+			placer: &T::AccountId,
+			asset_out: T::AssetCurrencyId,
+			asset_in: T::AssetCurrencyId,
+			nonce: T::Nonce,
+		) -> T::Hash {
+			(&placer, asset_in, asset_out, nonce).using_encoded(T::Hashing::hash)
+		}
+	}
```

Comment:
Nonce should not be enough to identify it?

Reply:
I was thinking that it would be better to have IDs that would not be easily determinable, but the `Nonce` would be usable for order identification.

Reply:
> it would be better to have IDs that would not be easily determinable

Why this? Can you expand?

Reply:
It was mostly in case there were anything potentially with predictable order ids, however I'm not aware of any situations where that would be an issue here.  I've updated to use nonce 👍